official/coin_build/man/man3/SoFontStyle.3
.TH "SoFontStyle" 3 "Sun May 28 2017" "Version 4.0.0a" "Coin" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SoFontStyle \- The \fBSoFontStyle\fP class changes the appearance of fonts for text rendering nodes\&.
.PP
Successive text rendering nodes will use fonts with the style settings of this node, if a font with the given settings can be found and loaded from the system\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Inventor/nodes/SoFontStyle\&.h>\fP
.PP
Inherits \fBSoFont\fP\&.
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBFamily\fP { \fBSERIF\fP, \fBSANS\fP, \fBTYPEWRITER\fP }"
.br
.ti -1c
.RI "enum \fBStyle\fP { \fBNONE\fP, \fBBOLD\fP, \fBITALIC\fP }"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "virtual \fBSoType\fP \fBgetTypeId\fP (void) const"
.br
.RI "Returns the type identification of an object derived from a class inheriting \fBSoBase\fP\&. This is used for run-time type checking and 'downward' casting\&. "
.ti -1c
.RI "\fBSoFontStyle\fP (void)"
.br
.ti -1c
.RI "\fBSbString\fP \fBgetFontName\fP (void) const"
.br
.ti -1c
.RI "virtual void \fBdoAction\fP (\fBSoAction\fP *action)"
.br
.ti -1c
.RI "virtual void \fBgetBoundingBox\fP (\fBSoGetBoundingBoxAction\fP *action)"
.br
.ti -1c
.RI "virtual void \fBGLRender\fP (\fBSoGLRenderAction\fP *action)"
.br
.ti -1c
.RI "virtual void \fBcallback\fP (\fBSoCallbackAction\fP *action)"
.br
.ti -1c
.RI "virtual void \fBpick\fP (\fBSoPickAction\fP *action)"
.br
.ti -1c
.RI "virtual void \fBgetPrimitiveCount\fP (\fBSoGetPrimitiveCountAction\fP *action)"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static \fBSoType\fP \fBgetClassTypeId\fP (void)"
.br
.ti -1c
.RI "static void \fBinitClass\fP (void)"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "\fBSoSFEnum\fP \fBfamily\fP"
.br
.ti -1c
.RI "\fBSoSFBitMask\fP \fBstyle\fP"
.br
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "virtual const \fBSoFieldData\fP * \fBgetFieldData\fP (void) const"
.br
.ti -1c
.RI "virtual \fB~SoFontStyle\fP ()"
.br
.in -1c
.SS "Static Protected Member Functions"

.in +1c
.ti -1c
.RI "static const \fBSoFieldData\fP ** \fBgetFieldDataPtr\fP (void)"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
The \fBSoFontStyle\fP class changes the appearance of fonts for text rendering nodes\&.
.PP
Successive text rendering nodes will use fonts with the style settings of this node, if a font with the given settings can be found and loaded from the system\&. 

If the specified font style cannot be found on the client system, a default fallback will be used\&. So the application programmer must consider \fBSoFontStyle\fP nodes as nodes giving \fIhints\fP about font settings, as you are \fInot\fP guaranteed to get exactly what you want\&.
.PP
Currently the SERIF family is mapped to 'Times New Roman', SANS is mapped to 'Arial' and TYPEWRITER is mapped to 'Courier New'\&.
.PP
Please note that this node is inheriting \fI\fBSoFont\fP\fP\&. Previous font settings in a branch will therefore be overwritten with the default values of \fI\fBSoFont\fP\fP if an empty \fI\fBSoFontStyle\fP\fP node is inserted in the branch\&.
.PP
\fINB!\fP This node is provided for reasons of compatibility with Open Inventor 2\&.1 but will in most cases result in the default font being rendered\&. It is highly recommended to use the \fBSoFont\fP node instead\&.
.PP
\fBFILE FORMAT/DEFAULTS:\fP 
.PP
.nf
FontStyle {
    name "defaultFont"
    size 10
    family SERIF
    style ()
}

.fi
.PP
.PP
\fBSince:\fP
.RS 4
Inventor 2\&.1 
.RE
.PP
\fBSee also:\fP
.RS 4
\fBSoFont\fP, \fBSoText2\fP, \fBSoText3\fP, \fBSoAsciiText\fP 
.RE
.PP

.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBSoFontStyle::Family\fP"
Enumeration of the font family to use\&. 
.SS "enum \fBSoFontStyle::Style\fP"
Enumeration of font style characteristics\&. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "SoFontStyle::SoFontStyle (void)"
Constructor\&. 
.SS "SoFontStyle::~SoFontStyle ()\fC [protected]\fP, \fC [virtual]\fP"
Destructor\&. 
.SH "Member Function Documentation"
.PP 
.SS "\fBSoType\fP SoFontStyle::getTypeId (void) const\fC [virtual]\fP"

.PP
Returns the type identification of an object derived from a class inheriting \fBSoBase\fP\&. This is used for run-time type checking and 'downward' casting\&. Usage example:
.PP
.PP
.nf
void foo(SoNode * node)
{
  if (node->getTypeId() == SoFile::getClassTypeId()) {
    SoFile * filenode = (SoFile *)node;  // safe downward cast, knows the type
  }
}
.fi
.PP
.PP
For application programmers wanting to extend the library with new nodes, engines, nodekits, draggers or others: this method needs to be overridden in \fIall\fP subclasses\&. This is typically done as part of setting up the full type system for extension classes, which is usually accomplished by using the pre-defined macros available through for instance \fBInventor/nodes/SoSubNode\&.h\fP (SO_NODE_INIT_CLASS and SO_NODE_CONSTRUCTOR for node classes), \fBInventor/engines/SoSubEngine\&.h\fP (for engine classes) and so on\&.
.PP
For more information on writing Coin extensions, see the class documentation of the toplevel superclasses for the various class groups\&. 
.PP
Reimplemented from \fBSoFont\fP\&.
.SS "const \fBSoFieldData\fP * SoFontStyle::getFieldData (void) const\fC [protected]\fP, \fC [virtual]\fP"
Returns a pointer to the class-wide field data storage object for this instance\&. If no fields are present, returns \fCNULL\fP\&. 
.PP
Reimplemented from \fBSoFont\fP\&.
.SS "\fBSbString\fP SoFontStyle::getFontName (void) const"
Returns a system-specific text string to use for font loading, based on the style settings of this node\&. 
.SS "void SoFontStyle::doAction (\fBSoAction\fP * action)\fC [virtual]\fP"
This function performs the typical operation of a node for any action\&. 
.PP
Reimplemented from \fBSoFont\fP\&.
.SS "void SoFontStyle::getBoundingBox (\fBSoGetBoundingBoxAction\fP * action)\fC [virtual]\fP"
Action method for the \fBSoGetBoundingBoxAction\fP\&.
.PP
Calculates bounding box and center coordinates for node and modifies the values of the \fIaction\fP to encompass the bounding box for this node and to shift the center point for the scene more towards the one for this node\&.
.PP
Nodes influencing how geometry nodes calculates their bounding box also overrides this method to change the relevant state variables\&. 
.PP
Reimplemented from \fBSoFont\fP\&.
.SS "void SoFontStyle::GLRender (\fBSoGLRenderAction\fP * action)\fC [virtual]\fP"
Action method for the \fBSoGLRenderAction\fP\&.
.PP
This is called during rendering traversals\&. Nodes influencing the rendering state in any way or who wants to throw geometry primitives at OpenGL overrides this method\&. 
.PP
Reimplemented from \fBSoFont\fP\&.
.SS "void SoFontStyle::callback (\fBSoCallbackAction\fP * action)\fC [virtual]\fP"
Action method for \fBSoCallbackAction\fP\&.
.PP
Simply updates the state according to how the node behaves for the render action, so the application programmer can use the \fBSoCallbackAction\fP for extracting information about the scene graph\&. 
.PP
Reimplemented from \fBSoFont\fP\&.
.SS "void SoFontStyle::pick (\fBSoPickAction\fP * action)\fC [virtual]\fP"
Action method for \fBSoPickAction\fP\&.
.PP
Does common processing for \fBSoPickAction\fP \fIaction\fP instances\&. 
.PP
Reimplemented from \fBSoFont\fP\&.
.SS "void SoFontStyle::getPrimitiveCount (\fBSoGetPrimitiveCountAction\fP * action)\fC [virtual]\fP"
Action method for the \fBSoGetPrimitiveCountAction\fP\&.
.PP
Calculates the number of triangle, line segment and point primitives for the node and adds these to the counters of the \fIaction\fP\&.
.PP
Nodes influencing how geometry nodes calculates their primitive count also overrides this method to change the relevant state variables\&. 
.PP
Reimplemented from \fBSoFont\fP\&.
.SH "Member Data Documentation"
.PP 
.SS "\fBSoSFEnum\fP SoFontStyle::family"
Font family hint\&. 
.SS "\fBSoSFBitMask\fP SoFontStyle::style"
Font style hint\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for Coin from the source code\&.


insanity/coin/src/nodes/SoFontStyle.cpp
/**************************************************************************\
 * Copyright (c) Kongsberg Oil & Gas Technologies AS
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 * 
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 * 
 * Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
\**************************************************************************/

/*!
  \class SoFontStyle SoFontStyle.h Inventor/nodes/SoFontStyle.h
  \brief The SoFontStyle class changes the appearance of fonts for text rendering nodes.
  \ingroup nodes

  Successive text rendering nodes will use fonts with the style
  settings of this node, if a font with the given settings can be
  found and loaded from the system.

  If the specified font style cannot be found on the client system, a
  default fallback will be used. So the application programmer must
  consider SoFontStyle nodes as nodes giving \e hints about font
  settings, as you are \e not guaranteed to get exactly what you want.

  Currently the SERIF family is mapped to "Times New Roman", SANS is
  mapped to "Arial" and TYPEWRITER is mapped to "Courier New".

  Please note that this node is inheriting \e SoFont. Previous font
  settings in a branch will therefore be overwritten with the default
  values of \e SoFont if an empty \e SoFontStyle node is inserted in
  the branch.

  \e NB! This node is provided for reasons of compatibility with Open
  Inventor 2.1 but will in most cases result in the default font being
  rendered. It is highly recommended to use the SoFont node instead.

  <b>FILE FORMAT/DEFAULTS:</b>
  \code
    FontStyle {
        name "defaultFont"
        size 10
        family SERIF
        style ()
    }
  \endcode

  \since Inventor 2.1
  \sa SoFont, SoText2, SoText3, SoAsciiText
*/

// *************************************************************************

#include <Inventor/nodes/SoFontStyle.h>

#include <Inventor/actions/SoGLRenderAction.h>
#include <Inventor/actions/SoGetBoundingBoxAction.h>
#include <Inventor/elements/SoFontNameElement.h>
#include <Inventor/elements/SoFontSizeElement.h>
#include <Inventor/elements/SoOverrideElement.h>
#include <Inventor/errors/SoDebugError.h>

#include "nodes/SoSubNodeP.h"

// *************************************************************************

/*!
  \enum SoFontStyle::Family
  Enumeration of the font family to use.
*/

/*!
  \enum SoFontStyle::Style
  Enumeration of font style characteristics.
*/

/*!
  \var SoSFEnum SoFontStyle::family
  Font family hint.
*/
/*!
  \var SoSFBitMask SoFontStyle::style
  Font style hint.
*/


// *************************************************************************

SO_NODE_SOURCE(SoFontStyle);

/*!
  Constructor.
*/
SoFontStyle::SoFontStyle(void)
{
  SO_NODE_INTERNAL_CONSTRUCTOR(SoFontStyle);

  SO_NODE_ADD_FIELD(family, (SoFontStyle::SERIF));
  SO_NODE_ADD_FIELD(style, (SoFontStyle::NONE));

  SO_NODE_DEFINE_ENUM_VALUE(Family, SERIF);
  SO_NODE_DEFINE_ENUM_VALUE(Family, SANS);
  SO_NODE_DEFINE_ENUM_VALUE(Family, TYPEWRITER);
  SO_NODE_SET_SF_ENUM_TYPE(family, Family);

  SO_NODE_DEFINE_ENUM_VALUE(Style, NONE);
  SO_NODE_DEFINE_ENUM_VALUE(Style, BOLD);
  SO_NODE_DEFINE_ENUM_VALUE(Style, ITALIC);
  SO_NODE_SET_SF_ENUM_TYPE(style, Style);
}

/*!
  Destructor.
*/
SoFontStyle::~SoFontStyle()
{
}

// Doc from superclass.
void
SoFontStyle::initClass(void)
{
  SO_NODE_INTERNAL_INIT_CLASS(SoFontStyle, SO_FROM_INVENTOR_2_1|SoNode::VRML1);
}

/*!
  Returns a system-specific text string to use for font loading, based
  on the style settings of this node.
*/
SbString
SoFontStyle::getFontName(void) const
{

  SbString fontname(this->name.getValue().getString());

#if COIN_DEBUG
  static bool messageflag = true;
  if (messageflag && (fontname != "defaultFont")) {
    SoDebugError::postWarning("SoFontStyle::getFontName",
                              "Font name ('%s') is ignored when using "
                              "FontStyle nodes. Use the 'family' and "
                              "'style' fields instead.", 
                              fontname.getString());    
    messageflag = false; // Only display this message once.
  }
#endif

  switch (this->family.getValue()) {
  case SoFontStyle::SERIF:
    fontname = "Times New Roman";
    break;
  case SoFontStyle::SANS:
    fontname = "Arial";
    break;
  case SoFontStyle::TYPEWRITER:
    fontname = "Courier New";
    break;
#if COIN_DEBUG
  default:
    SoDebugError::postWarning("SoFontStyle::getFontName",
                              "value of family field is invalid, setting to SERIF");
    fontname = "Times New Roman";
    break;
#endif // COIN_DEBUG
  }
  

  // If this doesn't hold up, we need to include a few more cases in
  // the switch block.
  assert(SoFontStyle::NONE == 0);

  switch (this->style.getValue()) {
  case SoFontStyle::NONE:
    break;
  case SoFontStyle::BOLD:
    fontname += ":Bold";
    break;
  case SoFontStyle::ITALIC:
    fontname += ":Italic";
    break;
  case (SoFontStyle::BOLD | SoFontStyle::ITALIC):
    fontname += ":Bold Italic";
    break;
#if COIN_DEBUG
  default:
    SoDebugError::postWarning("SoFontStyle::getFontName",
                              "value of style field is invalid");
    break;
#endif // COIN_DEUG
  }

  return fontname;
}

// Doc from superclass.
void
SoFontStyle::doAction(SoAction * action)
{
  
  SoState * state = action->getState();
  uint32_t flags = SoOverrideElement::getFlags(state);
  
#define TEST_OVERRIDE(bit) ((SoOverrideElement::bit & flags) != 0)
  
  if (!name.isIgnored() && !TEST_OVERRIDE(FONT_NAME)) {
    SoFontNameElement::set(action->getState(), this, this->getFontName());
    if (this->isOverride()) {
      SoOverrideElement::setFontNameOverride(state, this, true);
    }
  }
  if (!size.isIgnored() && !TEST_OVERRIDE(FONT_SIZE)) {
    SoFontSizeElement::set(action->getState(), this, this->size.getValue());
    if (this->isOverride()) {
      SoOverrideElement::setFontSizeOverride(state, this, true);
    }
  }

#undef TEST_OVERRIDE

}

// Doc from superclass.
void
SoFontStyle::getBoundingBox(SoGetBoundingBoxAction * action)
{
  this->doAction(action);
}

// Doc from superclass.
void
SoFontStyle::GLRender(SoGLRenderAction * action)
{
  this->doAction(action);
}

// Doc from superclass.
void
SoFontStyle::callback(SoCallbackAction * action)
{
  SoFontStyle::doAction((SoAction *)action);
}

// Doc from superclass.
void
SoFontStyle::pick(SoPickAction * action)
{
  SoFontStyle::doAction((SoAction *)action);
}

// Doc from superclass.
void
SoFontStyle::getPrimitiveCount(SoGetPrimitiveCountAction * action)
{
  SoFontStyle::doAction((SoAction *)action);
}
